actors/datacap/src/lib.rs

```diff
@@ -380,22 +380,37 @@ impl Actor {
 
 /// Implementation of the token library's messenger trait in terms of the built-in actors'
 /// runtime library.
-struct SyscallProvider<'a, RT> {
-    rt: &'a mut RT,
+#[repr(transparent)]
+struct SyscallProvider<RT> {
+    rt: RT,
 }
 
-impl<'a, RT> Syscalls for &SyscallProvider<'a, RT>
+impl<RT> SyscallProvider<RT> {
+    pub fn new(rt: &mut RT) -> &mut Self {
+        unsafe { &mut *(rt as *mut RT as *mut SyscallProvider<RT>) }
```

Comment:
I think you understand this kind of thing better than me, but I'm curious whether you think this is a _necessary_ construction, or an optimisation or sugar for something (it does seem to have made the use sites cleaner)? Ditto for implementing `Syscalls` for both the provider and reference-to-provider. 

What would need to be different (either in built-in actors or the token lib) for this to be simpler? One guess: `&mut self` in the `Syscalls` trait?

Reply:
`Token::wrap` now takes the actor runtime by reference_ That means the `ActorRuntime` needs to be rooted on the stack somehow.

The `ActorRuntime` takes the `SyscallProvider` by reference as well. This change means we don't need to explicitly store the `SyscallProvider` on the stack as well.

Without the this change, we'd need to:

1. Construct and store the syscall provider on the stack (as we were doing previously).
2. Construct and store the `ActorRuntime` on the stack (new).
3. Pass the reference to the `ActorRuntime` into `as_token` (as we're currently doing).

I.e.:

```rust
let msg = SyscallProvider { rt };
let art = actor_runtime(&msg); // NEW
let token = as_token(&mut st, &art);
```

Previously, the line marked as "NEW" lived in `as_token`, but that's no longer possible.

With this change, we can fold those first two lines into one:

```rust
let art = actor_runtime(rt);
let token = as_token(&mut st, &art);
```

Reply:
Note: I first tried to abstract over the by-reference nature of `SyscallProvider`. I.e., I tried to allow storing it in the `ActorRuntime` by value and/or by reference.

Unfortunately, that doesn't work because the blockstore borrows the runtime:

1. If I call `rt.blockstore()` first, the runtime's reference is borrowed and I can't move it into the `SyscallProvider`.
2. If I call `syscall_provider.rt.blockstore()` later, I can't move the `SyscallProvider`.

Reply:
Ok, so this boils down to sugar to turn a 3-line boilerplate into 2 (quite a few times).

Could you please leave a comment in the code similar to your explanation above, e.g. showing the alternative boilerplate? If I was coding it I would probably have just accepted the boilerplate and made a note to feed back API suggestions to Helix (I am guessing the compiled code is the same either way). I'm ok with this magic instead, but let's make it maximally accessible to whoever comes along next to maintain things here.

Reply:
Done.